### ISSUE TITLE
Tautomer canonicalization bug

### ISSUE THREAD

**Issue by msteijaert** [rdkit/rdkit#2363]
**Description:**

Great to see that MolVS is now integrated into rdkit!

As already mentioned in https://github.com/mcs07/MolVS/issues/31, there is a bug in the tautomer cannonicalization, which makes stereo centers disappear in some cases.

For example (rdkit version 2018.09.2):

```
from rdkit import Chem 
from rdkit.Chem.MolStandardize import standardize

for smiles in ["NC(=O)c1[nH]nc(c1O)[C@@H]2O[C@H](CO)[C@@H](O)[C@H]2O",
               "CC1=C(C(=O)C[C@@H]1OC(=O)[C@@H]2[C@H](C2(C)C)/C=C(\C)/C(=O)OC)C/C=C\C=C"
               "N[C@@H](C)C(=O)O","O1CCCC[C@@H]1C"]: 
    mol = Chem.MolFromSmiles(smiles)
    mol2 = standardize.Standardizer().tautomer_parent(mol)
    print('original smiles = ' + smiles + ' (' + str(len(Chem.FindMolChiralCenters(mol))) + ' chiral centers)')
    print('tautomer parent = ' + Chem.MolToSmiles(mol2) + ' (' + str(len(Chem.FindMolChiralCenters(mol2))) + ' chiral centers)')
    print('------------------------------')
```

output:
```
original smiles = NC(=O)c1[nH]nc(c1O)[C@@H]2O[C@H](CO)[C@@H](O)[C@H]2O (4 chiral centers)
tautomer parent = NC(=O)C1NNC(c2oc(CO)c(O)c2O)C1O (0 chiral centers)
------------------------------
original smiles = CC1=C(C(=O)C[C@@H]1OC(=O)[C@@H]2[C@H](C2(C)C)/C=C(\C)/C(=O)OC)C/C=C\C=CN[C@@H](C)C(=O)O (4 chiral centers)
tautomer parent = COC(=O)C(C)=CC1C(C(=O)OC2=C(C)C(C/C=C\CC=N[C@@H](C)C(=O)O)C(=O)C2)C1(C)C (1 chiral centers)
------------------------------
original smiles = O1CCCC[C@@H]1C (1 chiral centers)
tautomer parent = C[C@H]1CCCCO1 (1 chiral centers)
------------------------------
```

**Comment by greglandrum** [comment]
Confirmed.

**Comment by greglandrum** [comment]
This also came up during this discussion: https://www.mail-archive.com/rdkit-discuss@lists.sourceforge.net/msg09025.html

**Comment by Birdlet** [comment]
Hi, all.

Is this bug is still open?

I'm using this tautomerization function to prepare ligands before docking, and I notice this problem too.

When I checked the python source code, I found it __tautomerize a molecule round by round, like expanding a tree__, which made a combinationary explosion.

Besides, an empirical scoring function, which __highly prefer (carbocyclic) aromatic rings__, ordered `"NC(=O)C1NNC(c2oc(CO)c(O)c2O)C1O"` (carbocyclic aromatic ring) higher than original `"NC(=O)c1[nH]nc(c1O)[C@@H]2O[C@H](CO)[C@@H](O)[C@H]2O"` (heteroaromatic ring).

I have two solutions.

1. we can define a `depth`  parameter for cutting the tree hight, which means we only consider directly tautomerized molecules, eg., `depth` in (1, 2).
2. we can use `probablity_score` as a scoring function than a large number, and multiplies the `probablity_score` when `depth` increasing, which should decrease unreasonable tautomerization path.

If this suggestion sounds good, I think I can write some code for it.

**Comment by msteijaert** [comment]
@Birdlet. I think it's still open, as I still get the same behavior with version 2020.03.2. and the new `rdkit.Chem.MolStandardize.rdMolStandardize.TautomerEnumerator` interface.

However, judging from https://github.com/rdkit/rdkit/issues/2990 by @greglandrum, at least some of the stereo information is actually removed on purpose. Nevertheless, in the first smiles in my code there is still the issue of the "double bonds jumping from one ring to the next" as mentioned by @driesvr in https://github.com/mcs07/MolVS/issues/31.

**Comment by bp-kelley** [comment]
I agree with the loss of stereo generally, however, for a canonical tautomer, I expect a flavors where we can keep tetrahedral or double bond stereo would be useful.

**Comment by ptosco** [comment]
The new `CleanupParameters` `tautomerRemoveSp3Stereo` and `tautomerRemoveBondStereo` were added as part of https://github.com/rdkit/rdkit/pull/3327 and will be available starting from release 2020.09, i.e. in a couple of weeks.

**Comment by greglandrum** [comment]
I'm coming back to this after a few years further progress on the tautomer enumeration.
Here's the status using the 2024.03.1 version of the RDKit and the C++ version of tautomer enumeration/canonicalization.

The code:
```
from rdkit import Chem
from rdkit.Chem.MolStandardize import rdMolStandardize
tenum = rdMolStandardize.TautomerEnumerator()
for smiles in ["NC(=O)c1[nH]nc(c1O)[C@@H]2O[C@H](CO)[C@@H](O)[C@H]2O",
               "CC1=C(C(=O)C[C@@H]1OC(=O)[C@@H]2[C@H](C2(C)C)/C=C(\C)/C(=O)OC)C/C=C\C=C"
               "N[C@@H](C)C(=O)O","O1CCCC[C@@H]1C"]: 
    mol = Chem.MolFromSmiles(smiles)
    mol2 = rdMolStandardize.TautomerParent(mol)
    print('original smiles = ' + smiles + ' (' + str(len(Chem.FindMolChiralCenters(mol))) + ' chiral centers)')
    print('tautomer parent = ' + Chem.MolToSmiles(mol2) + ' (' + str(len(Chem.FindMolChiralCenters(mol2))) + ' chiral centers)')
    print('------------------------------')
```
Output:
```
original smiles = NC(=O)c1[nH]nc(c1O)[C@@H]2O[C@H](CO)[C@@H](O)[C@H]2O (4 chiral centers)
tautomer parent = NC(=O)c1[nH]nc([C@@H]2O[C@H](CO)[C@@H](O)[C@H]2O)c1O (4 chiral centers)
------------------------------
original smiles = CC1=C(C(=O)C[C@@H]1OC(=O)[C@@H]2[C@H](C2(C)C)/C=C(\C)/C(=O)OC)C/C=C\C=CN[C@@H](C)C(=O)O (4 chiral centers)
tautomer parent = COC(=O)/C(C)=C/[C@@H]1C(C(=O)O[C@H]2CC(=O)C(C/C=C\CC=NC(C)C(=O)O)=C2C)C1(C)C (2 chiral centers)
------------------------------
original smiles = O1CCCC[C@@H]1C (1 chiral centers)
tautomer parent = C[C@H]1CCCCO1 (1 chiral centers)
------------------------------
```
So with default parameters only the second molecule has stereo centers in a tautomeric region (this is what is required to lose stereochemistry).

The second molecule has two chiral centers removed, these are both involved in tautomeric systems.

Here's a schematic of the first of those:
```
In [34]: tenum = rdMolStandardize.TautomerEnumerator()

In [35]: m = Chem.MolFromSmiles('FC1C[C@H]1C(=O)OC')

In [36]: for taut in tenum.Enumerate(m):
    ...:     print(Chem.MolToSmiles(taut))
    ...: 
COC(=O)C1CC1F
COC(O)=C1CC1F
```
And the second:
```
In [37]: m = Chem.MolFromSmiles('CN[C@H](C)C(=O)O')

In [38]: for taut in tenum.Enumerate(m):
    ...:     print(Chem.MolToSmiles(taut))
    ...: 
CNC(C)=C(O)O
CNC(C)C(=O)O
```

Both of these cases have a tautomer which involves a double bond to the chiral center, so the chirality is lost. 

I'm going to close this issue, because I think the behavior is now what is expected. We can re-open it if there's disagreement about that.
